PostHog/Replay/NetworkRequestSample.swift

```diff
@@ -0,0 +1,257 @@
+        var responseBodyLength: Int?
+
+        var durationMs: UInt64?
+
+        var isProcessed: Bool = false
+
+        // called when a request starts loading
+        func start(request: URLRequest) {
+            requestStartTime = getMonotonicTimeInMilliseconds()
+            requestURL = request.url?.absoluteURL
+            requestMethod = request.httpMethod
+            requestHeaders = request.normalizedHeaderFields ?? [:]
+
+            // grab content-type. Keys normalized with .lowercase()
+            if let contentType = requestHeaders?["content-type"] as? String {
+                let contentType = contentType.components(separatedBy: ";")[0]
+                requestContentTypeRaw = contentType
+                requestContentType = ContentType(contentType: contentType)
+            }
+
+            // grab request body
+            if let requestData = request.httpBody ?? request.httpBodyStream?.consume() {
+                if let responseContentType, responseContentType == .image {
+                    // don't record response body for image types
+                    let bodyStr = requestData.base64EncodedString(options: .endLineWithLineFeed)
+                    requestBodyLength = bodyStr.count
+                } else if let utfString = String(data: requestData, encoding: String.Encoding.utf8) {
+                    requestBodyStr = utfString
+                    requestBodyLength = utfString.count
+                }
+            }
+        }
+
+        // called on stopLoading (request was cancelled)
+        func stop() {
+            durationMs = relative(getMonotonicTimeInMilliseconds(), to: requestStartTime)
+        }
+
+        // called on didCompleteWithError
+        func complete(response: URLResponse, error: Error?) {
+            let completedTime = getMonotonicTimeInMilliseconds()
+            responseEndTime = completedTime
+            responseStatus = (response as? HTTPURLResponse)?.statusCode
+            responseHeaders = response.normalizedHeaderFields ?? [:]
+            responseError = error?.localizedDescription
+
+            if let contentType = responseHeaders?["content-type"] as? String {
+                let contentType = contentType.components(separatedBy: ";")[0]
+                responseContentTypeRaw = contentType
+                responseContentType = ContentType(contentType: contentType)
+            }
+
+            durationMs = relative(completedTime, to: requestStartTime)
+
+            if let responseData = responseData as? Data {
+                if let responseContentType, responseContentType == .image {
+                    // don't record response body for image types
+                    let bodyStr = responseData.base64EncodedString(options: .endLineWithLineFeed)
+                    responseBodyLength = bodyStr.count
+                } else if let utfString = String(data: responseData, encoding: String.Encoding.utf8) {
+                    responseBodyStr = utfString
```

Comment:
also this https://posthog.com/docs/session-replay/network-recording#posthog-truncated-the-request-or-response-body

Reply:
and this https://posthog.com/docs/session-replay/network-recording#a-network-request-says-it-was-redacted

Reply:
and this https://posthog.com/docs/session-replay/network-recording#how-to-register-a-callback-to-inspect-and-redact-each-network-request
ok enough :D